azurelinuxagent/ga/self_update_version_updater.py

```diff
@@ -0,0 +1,181 @@
+
+        next_hotfix_time = get_next_process_time(self._last_attempted_self_update_time,
+                                                 conf.get_self_update_hotfix_frequency())
+        next_regular_time = get_next_process_time(self._last_attempted_self_update_time,
+                                                  conf.get_self_update_regular_frequency())
+
+        return next_hotfix_time, next_regular_time
+
+    def _is_new_agent_allowed_update(self):
+        """
+        This method ensure that update is allowed only once per (hotfix/Regular) upgrade frequency
+        """
+        now = datetime.datetime.now()
+        next_hotfix_time, next_regular_time = self._get_next_upgrade_times(now)
+        upgrade_type = self._get_agent_upgrade_type(self._version)
+
+        if (upgrade_type == SelfUpdateType.Hotfix and next_hotfix_time <= now) or (
+                upgrade_type == SelfUpdateType.Regular and next_regular_time <= now):
+            # Update the last upgrade check time even if no new agent is available for upgrade
+            self._last_attempted_self_update_time = now
+            return True
+        return False
+
+    def _should_agent_attempt_manifest_download(self):
+        """
+        The agent should attempt to download the manifest if
+        the agent has not attempted to download the manifest in the last 1 hour
+        If we allow update, we update the last attempted manifest download time
+        """
+        now = datetime.datetime.now()
+
+        if self._last_attempted_manifest_download_time != datetime.datetime.min:
+            next_attempt_time = self._last_attempted_manifest_download_time + datetime.timedelta(seconds=conf.get_agentupdate_frequency())
+        else:
+            next_attempt_time = now
+
+        if next_attempt_time > now:
+            return False
+        self._last_attempted_manifest_download_time = now
+        return True
+
+    def is_update_allowed_this_time(self):
+        """
+        Checks if we allowed download manifest as per manifest download frequency
+        """
+        if not self._should_agent_attempt_manifest_download():
+            return False
+        return True
+
+    def check_and_switch_updater_if_changed(self, agent_family, gs_id):
+        """
+        Checks if there is a new goal state and decide if we need to continue with self-update or switch to rsm update.
+        if vm is not enabled for RSM updates or agent not supports GA versioning then we continue with self update, otherwise we raise exception to switch to rsm update.
+        if isVersionFromRSM is missing but isVMEnabledForRSMUpgrades is present in the goal state, we ignore the update as we consider it as invalid goal state.
+        """
+        if self._gs_id != gs_id:
+            self._gs_id = gs_id
+            if conf.get_enable_ga_versioning() and agent_family.is_vm_enabled_for_rsm_upgrades:
```

Comment:
another method that is not protecting itself against None

Reply:
which None are you talking? is it is_vm_enabled_for_rsm_upgrades? That is intentional. In the self-update, I don't want to fail with error (stop the updates) if it's None rather I want to continue with self-update until flags showed up. Once the enabled flag is there, we should have other flag otherwise we don't switch to rsm-update.

Reply:
Are you checking for None on line 127? If looks like a check on a boolean. If you are checking against None, please do an explicit comparison against None. 

Taking advantage that None evalueates as False in a bool context is OK for small, one-single-use scripts, not so good for production code in a large code base.

Reply:
Having said that, see my comment above wrt allowing those properties to be None. Please don't. Encapsulate the use of invalid values (None) in a single method.